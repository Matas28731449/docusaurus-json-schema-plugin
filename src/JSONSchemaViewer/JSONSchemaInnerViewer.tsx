import React from "react"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  schema: JSONSchema7
}

function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <></>
}

export default JSONSchemaInnerViewer
