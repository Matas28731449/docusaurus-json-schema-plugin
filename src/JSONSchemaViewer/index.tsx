import React, { useEffect, useState } from "react"
import Layout from "@theme/Layout"
import Translate from "@docusaurus/Translate"
import AllOfMerger from "json-schema-merge-allof"
import { Resolver } from "@stoplight/json-ref-resolver"

import type { JSONSchema7 } from "json-schema"

import JSONSchemaViewer from "./JSONSchemaInnerViewer"

type ViewerProperties = {
  loadSchema: () => Promise<JSONSchema7>
}

// merged representation of allOf array of schemas
async function mergeAllOf(
  schema: JSONSchema7
): Promise<Omit<JSONSchema7, "allOf">> {
  const mergedSchema = AllOfMerger(schema, {
    // Technically valid, but only interested on fields with name, at least for now ...
    ignoreAdditionalProperties: true,
  })

  return Promise.resolve(mergedSchema as Omit<JSONSchema7, "allOf">)
}

function JSONSchemaEditor(props: ViewerProperties): JSX.Element {
  const [schema, setSchema] = useState(
    undefined as undefined | Omit<JSONSchema7, "allOf">
  )
  const [fetchError, setFetchError] = useState(undefined as undefined | Error)

  useEffect(() => {
    props
      .loadSchema()
      .then((userSchema) => new Resolver().resolve(userSchema))
      .then((resolvedSchema) => mergeAllOf(resolvedSchema.result))
      .then((schemaWithoutAllOf) => setSchema(schemaWithoutAllOf))
      .catch((err) => setFetchError(err))
  }, [])

  return (
    <Layout>
      {fetchError !== undefined && (
        <div>
          <p>
            <Translate
              values={{
                errorMessage: fetchError.message,
                id: "docusaurus-json-schema-viewer-plugin.error",
              }}
            >
              {"This component crashed because of error: {errorMessage}."}
            </Translate>
          </p>
        </div>
      )}
      {schema === undefined && (
        <div>
          <p>
            <Translate values={{ id: "docusaurus-json-schema-viewer-plugin.loading" }}>
              {"Loading schema ..."}
            </Translate>
          </p>
        </div>
      )}
      {schema !== undefined && <JSONSchemaViewer schema={schema} />}
    </Layout>
  )
}

export default JSONSchemaEditor
