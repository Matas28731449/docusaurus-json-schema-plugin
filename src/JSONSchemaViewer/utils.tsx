import React from "react";

import type { JSONSchema7 } from "json-schema";

// To keep the same type everywhere here
type cleanSchema = Omit<JSONSchema7, "allOf">;

