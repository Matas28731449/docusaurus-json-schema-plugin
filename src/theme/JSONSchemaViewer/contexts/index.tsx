export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
export * from "@theme/JSONSchemaViewer/contexts/schemaHierarchy"
