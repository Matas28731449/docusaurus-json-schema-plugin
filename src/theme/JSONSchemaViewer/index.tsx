import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"
import { JSVOptionsContextProvider } from "./contexts/index"

import type { JSONSchema } from "./types"
import type { JSVOptions } from "./contexts/index"
import type { IResolveOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  // The schema to use
  schema: unknown
  // To customize the ref resolving
  resolverOptions?: IResolveOpts
  // To customize the viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
  // Other properties (who knows ?)
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema
  // Options for viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema, viewerOptions } = props
  // Title of the schema, for user friendliness
  const title =
    typeof schema !== "boolean" && schema?.title !== undefined
      ? schema.title
      : "Schema"

  // state for provider
  const startingState: JSVOptions = {
    fullSchema: schema,
    // spread provided attributes
    ...viewerOptions,
  }

  return (
    <JSVOptionsContextProvider value={startingState}>
      <Collapsible
        summary={<strong>{title}</strong>}
        detailsProps={{
          open: true,
        }}
      >
        <CreateNodes schema={schema} />
      </Collapsible>
    </JSVOptionsContextProvider>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions, viewerOptions } = props

  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema
  )

  useEffect(() => {
    // Time to do the job
    new Resolver()
      .resolve(originalSchema, resolverOptions)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return (
      <JSONSchemaInnerViewer
        schema={resolvedSchema}
        viewerOptions={viewerOptions}
      />
    )
  }
}
