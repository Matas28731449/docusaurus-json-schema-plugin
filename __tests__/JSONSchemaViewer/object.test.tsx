import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"

const testcases: [string, JSONSchema][] = [
  [
    "simple",
    {
      type: "object",
      description: "A simple object",
    },
  ],
  [
    "minProperties",
    {
      type: "object",
      minProperties: 2,
    },
  ],
  [
    "maxProperties",
    {
      type: "object",
      maxProperties: 3,
    },
  ],
  [
    "minProperties & maxProperties",
    {
      type: "object",
      minProperties: 2,
      maxProperties: 3,
    },
  ],
  [
    "additionalProperties (false)",
    {
      type: "object",
      properties: {
        number: { type: "number" },
        street_name: { type: "string" },
        street_type: { enum: ["Street", "Avenue", "Boulevard"] },
      },
      additionalProperties: false,
    },
  ],
  [
    "additionalProperties (JSON Schema)",
    {
      type: "object",
      properties: {
        number: { type: "number" },
        street_name: { type: "string" },
        street_type: { enum: ["Street", "Avenue", "Boulevard"] },
      },
      additionalProperties: { type: "string" },
    },
  ],
  [
    "additionalProperties with patternProperties",
    {
      type: "object",
      properties: {
        builtin: { type: "number" },
      },
      patternProperties: {
        "^S_": { type: "string" },
        "^I_": { type: "integer" },
      },
      additionalProperties: { type: "string" },
    },
  ],
  [
    "patternProperties",
    {
      type: "object",
      patternProperties: {
        "^S_": { type: "string" },
        "^I_": { type: "integer" },
      },
    },
  ],
  [
    "properties",
    {
      type: "object",
      properties: {
        number: { type: "number" },
        street_name: { type: "string" },
        street_type: { enum: ["Street", "Avenue", "Boulevard"] },
      },
    },
  ],
  [
    "propertyNames",
    {
      type: "object",
      propertyNames: {
        pattern: "^[A-Za-z_][A-Za-z0-9_]*$",
      },
    },
  ],
  [
    "required",
    {
      type: "object",
      properties: {
        id: { type: "number", readOnly: true },
        name: { type: "string" },
        email: { type: "string" },
        address: { type: "string" },
        telephone: { type: "string", deprecated: true },
        pin_code: { type: "string", writeOnly: true },
      },
      required: ["name", "email"],
    },
  ],
  [
    "Unspecified type (object)",
    {
      minProperties: 1,
    },
  ],
  [
    "Unspecified required properties",
    {
      required: ["prop1", "prop2"],
    },
  ],
]

describe("JSONSchemaViewer - object", () => {
  test.each(testcases)("test %s", async (_title, fakeSchema) => {
    // render the component
    let root: ReactTestRenderer | undefined

    await act(async () => {
      root = create(<JSONSchemaViewer schema={fakeSchema} />)
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
