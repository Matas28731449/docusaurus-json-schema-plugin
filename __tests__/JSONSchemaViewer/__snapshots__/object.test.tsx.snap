// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - object test Unspecified required properties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                prop1
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                any
              </span>
               
              <strong>
                required
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  any
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  Always valid
                </div>
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                prop2
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                any
              </span>
               
              <strong>
                required
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  any
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  Always valid
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test Unspecified type (object) 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 1 propertie(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test additionalProperties (JSON Schema) 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                number
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_name
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_type
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    <strong>
                      Possible values :
                    </strong>
                     
                    <ul>
                      <li>
                        <code>
                          Street
                        </code>
                      </li>
                      <li>
                        <code>
                          Avenue
                        </code>
                      </li>
                      <li>
                        <code>
                          Boulevard
                        </code>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                property name*
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test additionalProperties (false) 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                number
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_name
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_type
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    <strong>
                      Possible values :
                    </strong>
                     
                    <ul>
                      <li>
                        <code>
                          Street
                        </code>
                      </li>
                      <li>
                        <code>
                          Avenue
                        </code>
                      </li>
                      <li>
                        <code>
                          Boulevard
                        </code>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          🚨 
          No extra propertie(s) are authorized in this object
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test additionalProperties with patternProperties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                builtin
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                ^S_
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                ^I_
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                integer
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  integer
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                property name*
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test maxProperties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &lt;= 3 propertie(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test minProperties & maxProperties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 2 propertie(s)
          </code>
           
          <strong>
            AND
          </strong>
           
          <code>
            &lt;= 3 propertie(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test minProperties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 2 propertie(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test patternProperties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                ^S_
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                ^I_
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                integer
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  integer
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test properties 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                number
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_name
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                street_type
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    <strong>
                      Possible values :
                    </strong>
                     
                    <ul>
                      <li>
                        <code>
                          Street
                        </code>
                      </li>
                      <li>
                        <code>
                          Avenue
                        </code>
                      </li>
                      <li>
                        <code>
                          Boulevard
                        </code>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test propertyNames 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                ^[A-Za-z_][A-Za-z0-9_]*$
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                any
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  any
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  Always valid
                </div>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test required 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                id
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
               
              <strong>
                readOnly
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    🚨 
                    read only
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                name
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
               
              <strong>
                required
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                email
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
               
              <strong>
                required
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                address
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                telephone
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
               
              <strong>
                deprecated
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    🚨 
                    Deprecated
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
        <li>
          <details
            open={false}
          >
            <summary>
              <strong>
                pin_code
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
               
              <strong>
                writeOnly
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    🚨 
                    write only
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test simple 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        A simple object
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test unevaluatedProperties = false 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          🚨 
          No extra propertie(s) are authorized in this object
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - object test unevaluatedProperties as not a boolean 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li>
          <details
            open={false}
          >
            <summary>
              <code>
                property name*
              </code>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  number
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;
